### PR TITLE
Added caching for client model validators

### PR DESCRIPTION
**Pull request by ajaybhargavb**
#3972 

@rynowak @javiercn

### CODE REVIEW

**Comment by javiercn** [comment]
:shipit: when @rynowak has taken a look at it. It looks good to me

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ClientValidatorCache.cs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            }
+
+            if (count == 0)
+            {
+                return EmptyArray;
+            }
+
+            var validators = new IClientModelValidator[count];
+            var clientValidatorIndex = 0;
+            for (int i = 0; i < items.Count; i++)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ClientValidatorCache.cs]: This is a cop out and it's totally weak :trollface: - this code was perfect before

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsClientModelValidatorProvider.cs [hunk 2]
```diff
@@ -66,21 +66,39 @@ public void GetValidators(ClientValidatorProviderContext context)
...
 
-            foreach (var attribute in context.ValidatorMetadata.OfType<ValidationAttribute>())
+            for (var i = 0; i < context.Results.Count; i++)
             {
+                var validatorItem = context.Results[i];
+                if (validatorItem.Validator != null)
+                {
+                    hasRequiredAttribute |= validatorItem.Validator is RequiredAttributeAdapter;
+                    continue;
+                }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsClientModelValidatorProvider.cs]: Why is this additional check needed?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsClientModelValidatorProvider.cs]: Is this needed because of partial caching? Please add a comment

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsClientModelValidatorProvider.cs [hunk 3]
```diff
@@ -66,21 +66,39 @@ public void GetValidators(ClientValidatorProviderContext context)
 
             var hasRequiredAttribute = false;
 
-            foreach (var attribute in context.ValidatorMetadata.OfType<ValidationAttribute>())
+            for (var i = 0; i < context.Results.Count; i++)
             {
+                var validatorItem = context.Results[i];
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsClientModelValidatorProvider.cs]: We should skip all items that already have `item.Validator` set.

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/NumericClientModelValidatorProvider.cs [hunk 4]
```diff
@@ -27,7 +27,11 @@ public void GetValidators(ClientValidatorProviderContext context)
                 typeToValidate == typeof(double) ||
                 typeToValidate == typeof(decimal))
             {
-                context.Validators.Add(new NumericClientModelValidator());
+                context.Results.Add(new ClientValidatorItem
+                {
+                    Validator = new NumericClientModelValidator(),
+                    IsReusable = true
+                });
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/NumericClientModelValidatorProvider.cs]: This need to still work when the validators are partially cached. This will just keep adding more and more instances of this validator as-is.

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsClientModelValidatorProviderTest.cs [hunk 5]
```diff
@@ -91,13 +93,19 @@ public void UnknownValidationAttribute_IsNotAddedAsValidator()
...
             // Act
             provider.GetValidators(providerContext);
 
             // Assert
-            Assert.Empty(providerContext.Validators);
+            var validatorItem = Assert.Single(providerContext.Results);
+            Assert.Null(validatorItem.Validator);
+        }
+
+        private IList<ClientValidatorItem> GetValidatorItems(ModelMetadata metadata)
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsClientModelValidatorProviderTest.cs]: We should add a new test for the behavior where we 'synthesize' a client validator for `metadata.IsRequired`.

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataTypeClientModelValidatorProviderTest.cs [hunk 6]
```diff
@@ -49,13 +51,18 @@ public void GetValidators_DoesNotGetsNumericValidator_ForUnsupportedTypes(Type m
...
 
             // Act
             provider.GetValidators(providerContext);
 
             // Assert
-            Assert.Empty(providerContext.Validators);
+            Assert.Empty(providerContext.Results);
+        }
+
+        private IList<ClientValidatorItem> GetValidatorItems(ModelMetadata metadata)
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataTypeClientModelValidatorProviderTest.cs]: Same thing, add a test that we don't create duplicate number validators

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ClientValidatorCache.cs (hunk 1 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ClientValidatorCache.cs]: @javiercn is unhappy with it. https://github.com/aspnet/Mvc/pull/3990#discussion_r50844194

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsClientModelValidatorProvider.cs (hunk 2 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsClientModelValidatorProvider.cs]: Yes. Will add.

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsClientModelValidatorProvider.cs (hunk 3 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsClientModelValidatorProvider.cs]: Not sure what you mean. The code already does that.

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/NumericClientModelValidatorProvider.cs (hunk 4 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/NumericClientModelValidatorProvider.cs]: Done.

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsClientModelValidatorProviderTest.cs (hunk 5 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsClientModelValidatorProviderTest.cs]: Added.

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataTypeClientModelValidatorProviderTest.cs (hunk 6 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataTypeClientModelValidatorProviderTest.cs]: Added.

**Comment by ajaybhargavb** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsClientModelValidatorProvider.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsClientModelValidatorProvider.cs]: Yeah, I see now. SOLID.

**Comment by rynowak** [comment]
:shipit:
